moto/utilities/Paginator.py

```diff
@@ -0,0 +1,139 @@
+
+    if original_function:
+        return pagination_decorator(original_function)
+
+    return pagination_decorator
+
+
+class Paginator(object):
+    def __init__(
+        self,
+        max_results=None,
+        max_results_default=None,
+        starting_token=None,
+        page_ending_range_keys=None,
+        param_values_to_check=None,
+        fail_on_invalid_token=True,
+    ):
+        self._max_results = max_results if max_results else max_results_default
+        self._starting_token = starting_token
+        self._page_ending_range_keys = page_ending_range_keys
+        self._param_values_to_check = param_values_to_check
+        self._fail_on_invalid_token = fail_on_invalid_token
+        self._token_encoder = TokenEncoder()
+        self._token_decoder = TokenDecoder()
+        self._param_checksum = self._calculate_parameter_checksum()
+        self._parsed_token = self._parse_starting_token()
+
+    def _parse_starting_token(self):
+        if self._starting_token is None:
+            return None
+        # The starting token is a dict passed as a base64 encoded string.
+        next_token = self._starting_token
+        try:
+            next_token = self._token_decoder.decode(next_token)
+        except (ValueError, TypeError, UnicodeDecodeError):
+            if self._fail_on_invalid_token:
+                raise InvalidToken("Invalid token")
+            return None
+        if next_token.get("parameterChecksum") != self._param_checksum:
+            raise InvalidToken(
+                "Input inconsistent with page token: {}".format(str(next_token))
+            )
+        return next_token
+
+    def _calculate_parameter_checksum(self):
+        if not self._param_values_to_check:
+            return None
+        return reduce(
+            lambda x, y: x ^ y,
+            [hash(item) for item in self._param_values_to_check.items()],
+        )
+
+    def _check_predicate(self, item):
+        if self._parsed_token is None:
+            return False
+        page_ending_range_key = self._parsed_token["pageEndingRangeKey"]
+        predicate_values = page_ending_range_key.split("|")
+        for (index, attr) in enumerate(self._page_ending_range_keys):
+            curr_val = item[attr] if type(item) == dict else getattr(item, attr, None)
+            if not curr_val == predicate_values[index]:
+                return False
+        return True
+
+    def _build_next_token(self, next_item):
+        token_dict = {}
+        if self._param_checksum:
+            token_dict["parameterChecksum"] = self._param_checksum
+        range_keys = []
+        for (index, attr) in enumerate(self._page_ending_range_keys):
+            if type(next_item) == dict:
+                range_keys.append(next_item[attr])
+            else:
+                range_keys.append(getattr(next_item, attr))
+        token_dict["pageEndingRangeKey"] = "|".join(range_keys)
+        return TokenEncoder().encode(token_dict)
```

Comment:
This should be `self._token_encoder.encode(token_dict)`.  (A minor oversight from back when I first submitted this code.)

Reply:
Fixed